src/Workspaces/Core/Portable/ObsoleteSymbol/AbstractObsoleteSymbolService.cs

```diff
@@ -0,0 +1,176 @@
+                        {
+                            if (trivia.HasStructure)
+                            {
+                                stack.Add(trivia.GetStructure()!);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        SyntaxToken ProcessNode(SemanticModel semanticModel, SyntaxNode node)
+        {
+            if (syntaxFacts.IsUsingAliasDirective(node))
+            {
+                syntaxFacts.GetPartsOfUsingAliasDirective(node, out _, out var aliasToken, out var name);
+                if (!aliasToken.Span.IsEmpty)
+                {
+                    // Use 'name.Parent' because VB can't resolve the declared symbol directly from 'node'
+                    var symbol = semanticModel.GetDeclaredSymbol(name.GetRequiredParent(), cancellationToken);
+                    if (IsSymbolObsolete(symbol))
+                        result.Add(aliasToken.Span);
+                }
+
+                return aliasToken;
+            }
+            else if (syntaxFacts.IsImplicitObjectCreationExpression(node))
+            {
+                syntaxFacts.GetPartsOfImplicitObjectCreationExpression(node, out var creationKeyword, out _, out _);
+                if (!creationKeyword.Span.IsEmpty)
+                {
+                    var symbol = semanticModel.GetSymbolInfo(node, cancellationToken).Symbol;
+                    if (IsSymbolObsolete(symbol) || IsSymbolObsolete(symbol?.ContainingType))
+                        result.Add(creationKeyword.Span);
+                }
+            }
+
+            return default;
+        }
+
+        void ProcessToken(SemanticModel semanticModel, SyntaxToken token)
+        {
+            if (syntaxFacts.IsIdentifier(token))
+            {
+                ProcessIdentifier(semanticModel, token);
+            }
+            else if (token.RawKind == _dimKeywordKind)
+            {
+                ProcessDimKeyword(result, semanticModel, token, cancellationToken);
+            }
+        }
+
+        void ProcessIdentifier(SemanticModel semanticModel, SyntaxToken token)
+        {
+            if (syntaxFacts.IsDeclaration(token.Parent))
+            {
+                var symbol = semanticModel.GetDeclaredSymbol(token.Parent, cancellationToken);
+                if (IsSymbolObsolete(symbol))
+                    result.Add(token.Span);
+            }
+            else
+            {
+                var symbol = semanticModel.GetSymbolInfo(token, cancellationToken).Symbol;
+                if (IsSymbolObsolete(symbol))
+                    result.Add(token.Span);
+            }
+        }
+    }
+
+    protected static bool IsSymbolObsolete([NotNullWhen(true)] ISymbol? symbol)
+    {
+        // Avoid infinite recursion. Iteration limit chosen arbitrarily; cases are generally expected to complete on
+        // the first iteration or fail completely.
```

Comment:
nit.  i do not think this is actually possible.  first, an alias can't target other aliases.  so in the alias case, we unwrap immediately to namespace or type.  and, for named types, the compiler never generates infinite types.  importantly, many many many other 'symbol walkers' would outright break in many other places if that happened.

Reply:
I'm not worried about valid code here, but rather for invalid code. e.g. something like a cycle in using aliases.

Reply:
so aliases cannot refer to each other in our symbol model.  An alias only every points to a single namespace or type.  And neither of those two can ever point to an alias.  Aliases are actually kind odd, and really shouldn't necessarily be a symbol in the first place (you can't get them with GetSymbolInfo for example, you can only get them with GetAliasInfo).  So you should think of them more like a *single pointer* that points into the real symbol model, and the real symbol model never points back at them.

So loops in them aren't possible if you follow the above guidelines (like using GetDeclaredSymbol(UsingDirective) to get the alias in the first place, and then only looking at its target.